frontend/javascripts/oxalis/model/volumetracing/volumelayer.js

```diff
@@ -434,13 +416,13 @@ class VolumeLayer {
       } else if (mode === VolumeToolEnum.BRUSH) {
         // we don't want to connect the last and the first circle with a rectangle
         if (i !== contourList.length - 1) {
-          const [xa, ya, xb, yb, xc, yc, xd, yd] = this.getRectangleBetweenCircles(p1, p2, radius, [
-            scaleX / this.activeResolution[dimIndices[0]],
-            scaleY / this.activeResolution[dimIndices[1]],
-          ]);
-          if (this.vector2DistanceWithScale(p1, p2, [scaleX, scaleY]) > 1.5 * radius) {
-            Drawing.fillRectangle(xa, ya, xb, yb, xc, yc, xd, yd, setMap);
-          }
+          // const [xa, ya, xb, yb, xc, yc, xd, yd] = this.getRectangleBetweenCircles(p1, p2, radius, [
+          //   scaleX / this.activeResolution[dimIndices[0]],
+          //   scaleY / this.activeResolution[dimIndices[1]],
+          // ]);
+          // if (this.vector2DistanceWithScale(p1, p2, [scaleX, scaleY]) > 1.5 * radius) {
+          //   Drawing.fillRectangle(xa, ya, xb, yb, xc, yc, xd, yd, setMap);
+          // }
```

Comment:
@daniel-wer At first I thought this code was wrong, since the rectangles should already have been drawn, but then I read that it's necessary for the implementation of the auto-fill-feature to work (since that code doesn't "know" about the rectangles which were drawn earlier?).
I'm a bit hesitant to keep this code here, as all rectangles are effectively drawn twice then. I wonder whether we should simply remove the auto-fill feature (especially, now that we have the paint-bucket-tool). What do you think? :see_no_evil:

Reply:
> At first I thought this code was wrong, since the rectangles should already have been drawn, but then I read that it's necessary for the implementation of the auto-fill-feature to work (since that code doesn't "know" about the rectangles which were drawn earlier?).

To explain the status quo: The key here is that only voxels that were drawn with _this_ brush stroke should affect whether an area is auto-filled or not. Therefore, the `contourList` which saves the coordinates of all brush events is used to recreate a map of the brushed voxels (and one cannot simply look up the voxels in the buckets). This map needs to mirror the brushed voxels exactly to allow for correct auto-fill which is why the circles and rectangles need to be redrawn from the `contourList`.

> I'm a bit hesitant to keep this code here, as all rectangles are effectively drawn twice then.

To avoid drawing twice, we could also remember the painted voxels for the current brush stroke (instead of only the coordinates of the brush events) and reuse these here for the auto-fill at the end of the brush stroke. I think you can estimate the best whether this would be easy/feasible with the current code, but I assume it's not that easy as we don't know the size of the map in advance :thinking:

> I wonder whether we should simply remove the auto-fill feature (especially, now that we have the paint-bucket-tool).

I probably wouldn't remove it for now, regardless of whether we change anything in this PR or did you notice performance problems related to the auto-fill?

Reply:
Thank you for the summary! As we discussed in person, the current implementation of this PR restricts the auto-fill-feature by only kicking in if the start and end of the drawn "path" are so that the first and last brush circle are touching. The downside is that closing a circle and releasing the mouse somewhere else (e.g., when correcting some "bumps" in the outline or just when being a bit imprecise), won't trigger the auto-fill. However, the UI for that could also be improved by highlighting the start point and the "will perform auto-fill"-state.
In general, I think that the auto-fill-feature is not super crucial and would like to choose the more performant route. In case of negative feedback, we can re-think this :)

> did you notice performance problems related to the auto-fill?

Yes, when brushing a rather large area, there was a noticeable lag for me when releasing the brush :/ This felt especially weird to me, since everything was already brushed (the auto-fill-feature didn't fill anything) and no additional computation was necessary.